test/flock_test.exs

```diff
@@ -0,0 +1,91 @@
+defmodule FlockTest do
+  use ExUnit.Case, async: false
+
+  alias Flock.{
+    MockDiscovery,
+    MockSupervisor,
+    MockCluster,
+    MockPool
+  }
+
+  test "It will populate the cluster on supervisor boot" do
+    node = {"localhost", 123}
+    :ok = MockDiscovery.update([node])
+
+    {:ok, supervisor} = MockSupervisor.start_link([])
+
+    verify_nodes_equal([node])
+
+    assert MockCluster.get_node(:dummy) == {:ok, node}
+
+    poolname = MockPool.poolname(node)
+
+    :poolboy.transaction(poolname, fn worker ->
+      assert GenServer.call(worker, :name) == :localhost_123
+    end)
+
+    Supervisor.stop(supervisor)
+  end
+
+  test "It will balance across multiple nodes" do
+    nodes = [{"localhost", 123}, {"localhost", 234}]
+    :ok = MockDiscovery.update(nodes)
+
+    {:ok, supervisor} = MockSupervisor.start_link([])
+
+    verify_nodes_equal(nodes)
+
+    for k <- 1..4 do
+      {:ok, node} = MockCluster.get_node(k)
+
+      MockPool.poolname(node)
+      |> :poolboy.transaction(fn worker ->
+        assert GenServer.call(worker, :name) == MockPool.nodename(node)
+      end)
+    end
+
+    Supervisor.stop(supervisor)
+  end
+
+  test "It will survive healthchecks" do
+    nodes = [{"localhost", 123}, {"localhost", 234}]
+    :ok = MockDiscovery.update(nodes)
+
+    {:ok, supervisor} = MockSupervisor.start_link([])
+
+    verify_nodes_present(nodes)
+
+    nodes = [{"localhost", 123}]
+
+    :ok = MockDiscovery.update(nodes)
+    send(MockCluster, :health_check)
+    :timer.sleep(100)
+
+    verify_nodes_equal(nodes)
+    verify_nodes_present(nodes)
+
+    nodes = [{"localhost", 567}, {"localhost", 234}]
+    :ok = MockDiscovery.update(nodes)
+    send(MockCluster, :health_check)
+    :timer.sleep(100)
+
+    verify_nodes_equal(nodes)
+    verify_nodes_present(nodes)
+
+    Supervisor.stop(supervisor)
```

Comment:
can you use https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#on_exit/2 here?

Reply:
I tried and it got screwy, but I'll do another spike on it